binary_search_tree/tree.py

```diff
@@ -1,55 +1,150 @@
+                    current = current.right
+            if parent.key > key:
+                parent.left = TreeNode(key,value)
+            else:
+                parent.right = TreeNode(key,value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+
+        current = self.root
+
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else: 
+                current = current.left
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_helper(self, current_node, values):
+        # base case
+        if not current_node:
+            return values
+
+        self.inorder_helper(current_node.left, values)
+        values.append({
+            "key": current_node.key,
+            "value": current_node.value
+        })
+
+        self.inorder_helper(current_node.right,values)
+        return values
+    
+    # Time Complexity: O(log n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        values = []
+        return self.inorder_helper(self.root, values)
+
+    def preorder_helper(self, current_node, values):
+        # base case
+        if not current_node:
+            return values
+
+        values.append({
+            "key": current_node.key,
+            "value": current_node.value
+        })
 
-    # Time Complexity: 
-    # Space Complexity:     
+        self.preorder_helper(current_node.left,values)
+        self.preorder_helper(current_node.right, values)
+        return values
+
+    # Time Complexity: O(logn n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        values = []
+        return self.preorder_helper(self.root, values)
+    
+    def postorder_helper(self, current_node, values):
+        if not current_node:
+            return values
+
+        self.postorder_helper(current_node.left, values)
+        self.postorder_helper(current_node.right, values)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        values.append({
+            "key": current_node.key,
+            "value":current_node.value
+        })
+
+        return values
+    
+    # Time Complexity: O(log n) 
+    # Space Complexity: O(n)
     def postorder(self):
-        pass
+        values = []
+        return self.postorder_helper(self.root, values)
+
+    def height_helper(self, current_node):
+        if not current_node:
+            return 0
+        left = self.height_helper(current_node.left)
+        right = self.height_helper(current_node.right)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        return max(left, right) + 1
+    
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n) >> ?     
     def height(self):
```

Comment:
👍 , but the time complexity is O(n) because you hit every single node. The recursive depth is Log n for a balanced tree so the space complexity is correct.